### PR TITLE
fix: toggling initialized values from multi selections 

### LINKED ISSUE #5199: can't toggle initialized multi selection until it's cleared?
I'm using vega-lite and react-vega for a biggish new project and have quite a few questions and suggestions. I'll start with the most straightforward one to warm up.

I'm using vega-lite 3.3. It's a dashboard with a server-side component. Think of it as a generalization on Nanocubes and crossfilter (and the earlier iPlots). It will be open source but it's not released yet.

The charts update each other a lot, and they go outside of Vega to get new data. This means the selection must be preserved and re-applied with whatever has changed about the data and spec.

Selection initialization with `init` works, but I find that any initialized selection does not toggle, whether or not I explicitly set `"toggle": "event.shiftKey"`. Shift-clicking something new will add it to the selection, but shift-clicking within the selection will not remove anything.

If I clear the selection, this enables toggling for that selection.

Example below with two selections. The behavior is the same with one selection but this makes it really obvious because you can shift-alt-click to toggle the purple selection, but you can only add to the orange selection with shift-click, not remove, until the selection is cleared.

(please excuse the content, I'm using Chicago Crimes as my test data)

I'm eager to contribute once I find my way around!

[vega editor link](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAmCGAutIC4yghSA3WAbArgKYDOKYA2uBhOtRpPAJ4AOhZkAcgIIBKAwgHkAKgEk+AZUgAaKnSjwAlgFs2qSACYADAEYAzAFpt6-TqGaAbMk2armgHTXNALWmy6kAMYB7fADt4ZNqaACwArG4QAL4ycrRyDCyqUHw8IgCyItwAMmAAIlxpXADiAKKucpiKKuxaeobGpha2tg7WLjEVnj7+ZAAc6roRYNFDce5MrOxCABIlAGJC5Z1VSRo6BkYm2maW1i2O7UOY3n4BqNoAnL3BQyOxR-KJ7ABCXEJCJTwAmkvxKzXrepbHbNaytZy-dwnHrnUK9bS3DrUMbUBKTNTPACqPCKWV4PyR42Uq1qGwa2yaezBB0hqOhZzAoTCiNGDzRqzSwgEPDAADUStMxFkSmAZvNFoTUf81KSgY1djZqW1afR6WRdMELiz7hVHuioAI5nMShxxCLMryBFleZkimA+IKsrkeCaVZViQC6pt5aD7DTJarugyAOw3OR3Ogo+gTVa5Ep8EoABVE-LAiZ4XD4ogTbvkHplgO9FIV+2VAeOQbIwXU4XDAajlSeah4Amez0+BLZ0qgsqLIKpfrLbLVqF0oVrdAjyK7TagXHE4i4mKyEq7+Z7hfJ-cVg4h5agI7AYXM2sjM-1kGEsx5huNprK+4Y67WXq3lJ34MOuq6pzI5k0wantO34xuwKTpJkXA5EILriIm86SI+3YvmSwLvqWe7DpWqA1r0QEYA2eqrAA6iUXCJgIpp8iIVpvDRHC5k+1QFq+aElkqmHfoevSAXWrIgbOkAANIiLk3CJomtqMchvZvuxu5fp0h5GPhNDnqsiaYs8WRiGmZEJtRtGiJR0nPrJbG+p+jHcWGk71up7CZB8PDGi6HAGcRIgzGmWk6XwYC3mInymcxG6sT6A5WY+h7BAifE6ssgngWkYBmgAGpiUFgAhS4riFJKbhZkX+lhv44bxdn8YlF46QAipi3JgHixGGXixkMUhZmFRFH4lVx2FgEEqloA5ajTAIGR8KJD5rqFKFysWll9UpA16MNhHsuw4hCFBIkcEU+WeqhPUYYp8SHhO1BTgRo1QOlAVGiaZqHSxx2LcVQ79WVYBaOtt2QOmAjbV5mLtS9YVvdup3WQNwS2Vd9kCRevDiCZnVzeZJ0cWdUKrXFlUJX8glORB+Rg+jBXhe9vWfSt32XRg11qUjqxFAU2lSRTR0LVD2Mw99+NXW4AC6VARjAJAeAATgozCKF4vjsAIUvQAovh4GAHgABawFLAQdJAhC+N4qu+AA5mQYyQAAHpbQyQAAZgohC4NA7BqgGm1qAAjvgsD+AoiCKFgbBuNdkCMHb8ROy7btqKBnugWoXgq2reCQGHhJdLgKdR+4OAEKsZtS4Qkee94viq-LiuoJQuqEZgxAu4QHjV+wTu4PAhBS9ouaYAXRDsAAxNAFyhA7mjqBnupMzduqN83rcKAr7cKJ33dT-u-d4IPahD8GoTBpoABGujTxUkRDKLQtRAbSi6wA1uwx+60skBN7gLdt6gwBgG4jtry7j3POUpBJKHwJ3BQKoGBeDNmbT+7BCAh38HYYgWsFAO3gEJUu0DCBKDlpHNQvgFZsE9ivNQSgfBN2gF4AA7r4cgABCJBRt4B2DwFg0uwtoEx1dqQWuDkr6ezVoHEB4xZx13npAZKkEcj5EKKUPuUBuD8GEGIRCDxMBXk+A9O8z0HjX0nJnf+HcgFTx-vbJOUBwGQOgfAWB8DVgsJQWgjBnCy72zwQQ9gxDfCkPtuQ6xVDCA0PoeQZxbCOHYMYNwz2vDoD8IoIIuQhiohixAJEIAA)
```json
{
  "data": {
    "values": [
      {
        "type": "NARCOTICS",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 1045
      },
      {
        "type": "CRIMINAL DAMAGE",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 823
      },
      {
        "type": "THEFT",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 1984
      },
      {
        "type": "BATTERY",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 1581
      },
      {
        "type": "BURGLARY",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 545
      },
      {
        "type": "MOTOR VEHICLE THEFT",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 349
      },
      {
        "type": "OFFENSE INVOLVING CHILDREN",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 74
      },
      {
        "type": "DECEPTIVE PRACTICE",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 425
      },
      {
        "type": "ROBBERY",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 355
      },
      {
        "type": "ASSAULT",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 456
      },
      {
        "type": "OTHER OFFENSE",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 607
      },
      {
        "type": "CRIMINAL TRESPASS",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 258
      },
      {
        "type": "WEAPONS VIOLATION",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 87
      },
      {
        "type": "KIDNAPPING",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 12
      },
      {
        "type": "PUBLIC PEACE VIOLATION",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 84
      },
      {
        "type": "INTERFERENCE WITH PUBLIC OFFICER",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 41
      },
      {
        "type": "CRIM SEXUAL ASSAULT",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 27
      },
      {
        "type": "LIQUOR LAW VIOLATION",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 10
      },
      {
        "type": "HOMICIDE",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 13
      },
      {
        "type": "STALKING",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 5
      },
      {
        "type": "SEX OFFENSE",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 20
      },
      {
        "type": "PROSTITUTION",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 44
      },
      {
        "type": "ARSON",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 11
      },
      {
        "type": "INTIMIDATION",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 5
      },
      {
        "type": "GAMBLING",
        "time": "2013-12-01T06:00:00.000Z",
        "count": 1
      }
    ]
  },
  "description": "Ordinal chart",
  "encoding": {
    "x": {
      "field": "count",
      "type": "quantitative"
    },
    "y": {
      "field": "type",
      "type": "ordinal"
    },
    "color": {
      "value": "grey",
      "condition": [
        {
          "selection": "filter1",
          "value": "#d95f02"
        },
        {
          "selection": "filter2",
          "value": "#7570b3"
        }
      ]
    }
  },
  "mark": "bar",
  "selection": {
    "filter1": {
      "type": "multi",
      "toggle": "event.shiftKey",
      "empty": "none",
      "on": "mousedown[!event.altKey]",
      "fields": [
        "type"
      ],
      "init": {
        "type": [
          "CRIMINAL DAMAGE",
          "NARCOTICS",
          "OTHER OFFENSE"
        ]
      }
    },
    "filter2": {
      "type": "multi",
      "toggle": "event.shiftKey",
      "empty": "none",
      "on": "mousedown[event.altKey]",
      "fields": [
        "type"
      ]
    }
  }
}
```

### PR DESCRIPTION
**Pull request by arvind**
Fixes #5199, which revealed several minor bugs:

1. Now that initial values for selections are embedded directly as dataset values (rather than embedded in signal expressions), we no longer need to string escape model names. 

2. Datetime expressions should evaluate to numerical timestamps, rather than strings. 

3. Vega's `modify` expression should also test for object equality (which wasn't previously necessary for our logic because the same object was being reused in every tuple and thus strict equality testing `===` would correctly evaluate to `true`). vega/vega#1909.

4. ~~Although the typings/schema disallow it (and I believe should continue to disallow it, for simplicity/disambiguity), when a multi selection is projected over a single field, it is reasonable to expect people to use a short form initialization as seen in #5199. i.e., `{init: {Origin: ['Japan', 'USA']}}`. Now, we test for this one corner case, and expand it to what it should be `{init: [{Origin: 'Japan'}, {Origin: 'USA'}]}`~~

### CODE REVIEW

**Comment by arvind** [comment]
Marking this as a WIP -- after reading #4766, I want to consider whether there's a straightforward way of supporting columnar initialization of multi-selections which is probably the more correct fix than the awkward workaround this PR currently proposes. 

What makes columnar support slightly tricky is that to initialize ranged fields (e.g., binned fields), users specify an array of values. Thus, we need to distinguish `"init": {"x": [0, 100], "Origin": "Europe"}` (which should initialize just a single tuple because `x` is a ranged field) from `"init": {"x": [0, 100], "Origin": ["Europe", "Japan"]}` which should generate two tuples `(0, Europe), (100, Japan)`. Perhaps we test to ensure all arrayed properties have the same length?

**Comment by domoritz** [comment]
What if x and y are ranged fields? `"init": {"x": [0,20], "y": [10,30]}`? Does this mean `[[0,10],[20,30]]` or `[[0,20],[10,30]]`?

**Comment by gordonwoodhull** [comment]
Aha! Makes sense - confirmed, it does work if I use long form.

I guess I got the short form from the selection event, tried it, it seemed to work except for this tiny glitch, didn't think I might have the wrong format.

I validate with ajv every time I render, but I find the errors difficult to interpret. I have a few cases where things seem to work but validation fails. At some point I'll take the time to figure out whether vega-lite is being more permissive than the schema, or if I'm just passing extra/irrelevant/incomplete bits of spec due to generation.

**Comment by arvind** [comment]
> What if x and y are ranged fields?

If both `x` and `y` are ranged fields, then that `init` value would just be transformed trivially to `"init": [{"x": [0,20], "y": [10,30]}]` and stored as `"fields":  ["x", "y"], "values": [[0, 20], [10, 30]]"`.

**Comment by arvind** [comment]
I've removed the `WIP` label and dropped the commit related to columnar/shortform initialization of multi selections (ae7f0d4ae88593a35bd7bb9b4c11194564052f70) to keep this PR focused on bug fixes and unblock it for review. 

Columnar initialization is proving to be slightly trickier than I anticipated, and I exceeded my time box for it (it doesn't seem to be a sufficiently high priority for now given we support long form initialization). We're tracking this as an enhancement in #4766.

**Comment by arvind** [comment]
@gordonwoodhull thanks for confirming things work with long form! Agreed on the fact that the schema can sometimes throw confusing errors, and I think this is especially true for how selections can be initialized. I think this is where we can do better with additional documentation.

File: src/compile/selection/index.ts [hunk 1]
```diff
@@ -87,8 +87,8 @@ function getFacetModel(model: Model): FacetModel {
   return parent as FacetModel;
 }
 
-export function unitName(model: Model) {
-  let name = stringValue(model.name);
+export function unitName(model: Model, escape = true) {
```
Comment (domoritz) [inline src/compile/selection/index.ts]: We prefer to use objects for boolean properties to make them more understandable. 

```suggestion
export function unitName(model: Model, {escape} = {escape: true}) {
```

Comment (arvind) [reply src/compile/selection/index.ts]: Thanks for the suggestion, @domoritz! Implemented in 283419f.
